datafusion/physical-expr/src/array_expressions.rs

```diff
@@ -56,20 +67,29 @@ macro_rules! new_builder {
 
 macro_rules! array {
     ($ARGS:expr, $ARRAY_TYPE:ident, $BUILDER_TYPE:ident) => {{
-        // downcast all arguments to their common format
-        let args =
-            downcast_vec!($ARGS, $ARRAY_TYPE).collect::<Result<Vec<&$ARRAY_TYPE>>>()?;
-
-        let builder = new_builder!($BUILDER_TYPE, args[0].len());
+        let builder = new_builder!($BUILDER_TYPE, $ARGS[0].len());
         let mut builder =
-            ListBuilder::<$BUILDER_TYPE>::with_capacity(builder, args.len());
+            ListBuilder::<$BUILDER_TYPE>::with_capacity(builder, $ARGS.len());
+
         // for each entry in the array
-        for index in 0..args[0].len() {
-            for arg in &args {
-                if arg.is_null(index) {
-                    builder.values().append_null();
-                } else {
-                    builder.values().append_value(arg.value(index));
+        for index in 0..$ARGS[0].len() {
```

Comment:
I am not sure about this approach of taking either a `ListArray` or a `NullArray`

In the other functions, the way NULL is treated is that the input types are always the same (in this case ListArray) and the values would be `null` (aka `array.is_valid(i)` would return false for rows that are null.

Complicating matters is if you type a literal `null` in sql like:

```sql
select array_concat([1,2], null)
```

That comes to DataFusion as a `null` literal (with DataType::Null). The coercion / casting logic normally will coerce this to the appropriate type. 

For example, here is how I think arithmetic works with null:

```sql
select 1 + NULL
```

Arrives like 
```sql
ScalarValue::Int32(Some(1)) + ScalarValue::Null
```

And then the coercion logic will add a cast to Int32:

```sql
ScalarValue::Int32(Some(1)) + CAST(ScalarValue::Null, DataType::Int32)
```

And then the constant folder will collapse this into:


```sql
ScalarValue::Int32(Some(1)) + ScalarValue::Int32(None)
```

So by the time the arithmetic kernel sees it, it only has to deal with arguments of `Int32`

Reply:
I would expect that the array contains nulls values 

Like
```
select make_array(1, Null, 2 Null);
---- 
1, NULL, 2, NULL
```

Where the element of the ListArray are marked as null 🤔